### PR TITLE
OCLOMRS-606: Logging out on one tab should log the user out in all other tabs and windows

### PR DESCRIPTION
**Pull request by danuluma**
# JIRA TICKET NAME:
[Logging out on one tab should log the user out in all other tabs and windows](https://issues.openmrs.org/browse/OCLOMRS-606)

# Summary:
When one logs out on one tab, they are not logged out from other tabs and can still try to perform actions that require authentication. This fails and sometimes causes the app to crash.

This ticket is to fix this and make it such that when one logs outs, they are automatically logged out from all the other windows and tabs automatically. Also, when one logs in, they'll be automatically logged in to any open windows and/or tabs and be able to make authenticated requests.

### CODE REVIEW

File: src/tests/Dashboard/action/dictionaryAction.test.js [hunk 1]
```diff
@@ -185,6 +187,21 @@ describe('Test suite for dictionary actions', () => {
     });
   });
 
+  it('should alert a user and cancel the request if not logged in', () => {
+    const notifyMock = jest.fn();
+    notify.show = notifyMock;
+    localStorage.clear();
+    moxios.wait(() => {
+      moxios.requests.mostRecent();
```
Comment (Karuhanga) [inline src/tests/Dashboard/action/dictionaryAction.test.js]: Are we doing anything with this request?

File: src/components/Auth/index.jsx [hunk 2]
```diff
@@ -1,12 +1,29 @@
...
 
-export const mapStateToProps = ({ users: { loggedIn } }) => ({
-  loggedIn,
-});
+export const checkAuth = (event) => {
+  if (event.key === null) {
+    const token = localStorage.getItem('token');
+    if (!token) {
+      logoutAction();
+      history.go(0);
```
[the remark this answers is not in the document]
Comment (danuluma) [reply src/components/Auth/index.jsx]: `history.go(0)` simply reloads the current page, which is the behaviour intended. Its similar to plain js `window.location.reload()`

The `loggedIn` status is stored in the redux store,,which is **not** shared across tabs and windows. Logging out in one window does not change this value in other windows and tabs hence cannot be relied upon

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 1 above)
Comment (danuluma) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: Nothing, it's redundant. Fixed,, thank you

File: src/components/Auth/index.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Karuhanga) [reply src/components/Auth/index.jsx]: I think `history.go(0)` goes to the first index in the history, which might not be the current page.

[the remark this answers is not in the document]
Comment (Karuhanga) [reply src/components/Auth/index.jsx]: Anyway, what I'm wondering is, isn't dispatching the logged out action enough?

[the remark this answers is not in the document]
Comment (danuluma) [reply src/components/Auth/index.jsx]: `history.go(0)` goes to the first index in the history array. This is **always** the current page.....FYI positive numbers will go forward and negative numbers backwards, relative to the current location (0)

**Comment by danuluma** [comment]
The drop in test coverage has been addressed in #499

File: src/redux/reducers/authReducers.js [hunk 3]
```diff
@@ -30,14 +29,13 @@ export default (state = initialState, action) => {
         loading: false,
       };
     case AUTHENTICATION_FAILED:
```
Comment (Karuhanga) [inline src/redux/reducers/authReducers.js]: When authentication fails, does this mean we don't change state?

Comment (danuluma) [reply src/redux/reducers/authReducers.js]: Yes we are still changing the state same way as before,,

Comment (danuluma) [reply src/redux/reducers/authReducers.js]: Same as logging out,,just tied these two together

Comment (Karuhanga) [reply src/redux/reducers/authReducers.js]: Ahh. Missed that. Cool

**Pull request merged**